// package-json-validator/lib/utils/validatePeople.js
import { emailFormat, urlFormat } from "../formats.js";

//#region src/utils/validatePeople.ts
const isPersonArray = (obj) => {
	return Array.isArray(obj) && obj.every((item) => isPerson(item));
};
const isPerson = (obj) => {
	return typeof obj === "object" && obj !== null && "name" in obj;
};
/**
* Validate 'people' fields, which can be an object like this:
*
* {
*   "name" : "Barney Rubble"
*   "email" : "b@rubble.com",
*   "url" : "http://barnyrubble.tumblr.com/"
* }
*
* Or a single string like this:
* "Barney Rubble &lt;b@rubble.com> (http://barnyrubble.tumblr.com/)
* Or an array of either of the above.
*/
const validatePeople = (name, obj) => {
	const errors = [];
	function validatePerson(obj$1) {
		if (typeof obj$1 == "string") {
			const authorFields = /^([^<(\s][^<(]*)?(\s*<(.*?)>)?(\s*\((.*?)\))?/.exec(obj$1);
			if (authorFields) {
				const authorName = authorFields[1];
				const authorEmail = authorFields[3];
				const authorUrl = authorFields[5];
				validatePerson({
					email: authorEmail,
					name: authorName,
					url: authorUrl
				});
			} else errors.push(`Unable to parse person string: ${obj$1}`);
		} else if (typeof obj$1 == "object") {
			if (!obj$1.name) errors.push(`${name} field should have name`);
			if (obj$1.email && !emailFormat.test(obj$1.email)) errors.push(`Email not valid for ${name}: ${obj$1.email}`);
			if (obj$1.url && !urlFormat.test(obj$1.url)) errors.push(`URL not valid for ${name}: ${obj$1.url}`);
			if (obj$1.web && !urlFormat.test(obj$1.web)) errors.push(`URL not valid for ${name}: ${obj$1.web}`);
		} else errors.push("People field must be an object or a string");
	}
	if (obj instanceof Array) for (let i = 0; i < obj.length; i++) validatePerson(obj[i]);
	else validatePerson(obj);
	return errors;
};

//#endregion
export { isPerson, isPersonArray, validatePeople };
