// rcc-config-core/interpreter/provider-defaults.js
// Provider defaults and capabilities for config-core
// NOTE: 初期仅覆盖 openai / responses / anthropic，以及基于 openai 协议族的内置家族

/**
 * Provider 默认行为定义
 *
 * 约定:
 * - key 为 provider family 名称，例如 'openai' / 'responses'
 * - 仅在 canonical builder 中作为“系统默认”，可被 userConfig 覆盖
 */

export const PROVIDER_DEFAULTS = {
  openai: {
    id: 'openai',
    providerType: 'openai',
    protocol: 'openai-chat',
    baseUrl: 'https://api.openai.com/v1',
    endpoint: '/chat/completions',
    // 不在 defaults 中声明模型，强制用户在配置中显式指定，避免隐式回退
    // defaultModel: 'gpt-4.1-mini',
    timeoutMs: 300000,
    maxRetries: 3,
    headers: {
      'Content-Type': 'application/json'
    },
    auth: {
      required: ['apikey'],
      optional: ['oauth']
    }
  },
  // Qwen: OpenAI 协议族，默认走 Qwen 兼容模式，由 compatibility 层处理差异。
  // 协议族仍为 openai，providerId=Qwen 仅用于路由与兼容层。
  qwen: {
    id: 'qwen',
    providerType: 'openai',
    protocol: 'openai-chat',
    baseUrl: 'https://portal.qwen.ai/v1',
    endpoint: '/chat/completions',
    timeoutMs: 300000,
    maxRetries: 3,
    headers: {
      'Content-Type': 'application/json'
    },
    // 优先通过 OAuth 登录（与 rcc CLI 行为一致），允许显式 apiKey 覆盖。
    auth: {
      required: ['oauth'],
      optional: ['apikey']
    }
  },
  responses: {
    id: 'responses',
    providerType: 'responses',
    protocol: 'openai-responses',
    baseUrl: 'https://api.openai.com/v1',
    endpoint: '/responses',
    defaultModel: 'gpt-4.1-mini',
    timeoutMs: 300000,
    maxRetries: 3,
    headers: {
      'Content-Type': 'application/json'
    },
    auth: {
      required: ['apikey'],
      optional: []
    }
  },
  // iFlow: OpenAI Chat 协议族，HTTP 行为由 openai-http-provider 实现，兼容差异交给 iflow-compatibility。
  iflow: {
    id: 'iflow',
    providerType: 'openai',
    protocol: 'openai-chat',
    baseUrl: 'https://apis.iflow.cn/v1',
    endpoint: '/chat/completions',
    timeoutMs: 300000,
    maxRetries: 3,
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': 'iFlow-Cli'
    },
    // 也是以 OAuth 为主，允许 apiKey 直连。
    auth: {
      required: ['oauth'],
      optional: ['apikey']
    }
  },
  anthropic: {
    id: 'anthropic',
    providerType: 'anthropic',
    // 出站协议按 llmswitch 侧定义，这里只声明 messages 协议族
    protocol: 'anthropic-messages',
    baseUrl: 'https://api.anthropic.com/v1',
    endpoint: '/messages',
    defaultModel: 'claude-3.5-sonnet-latest',
    timeoutMs: 300000,
    maxRetries: 3,
    headers: {
      'Content-Type': 'application/json'
    },
    auth: {
      required: ['apikey'],
      optional: []
    }
  }
};
